File: android/app/src/main/kotlin/com/danieldallos/pigeon_multi_instance_demo/NativeView.kt
package com.danieldallos.pigeon_multi_instance_demo

import android.content.Context
import android.graphics.Color
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.platform.PlatformView
import com.danieldallos.pigeon_multi_instance_demo.pigeon.NativeTextApi
import com.danieldallos.pigeon_multi_instance_demo.pigeon.FlutterTextApiHandler
import com.danieldallos.pigeon_multi_instance_demo.pigeon.PigeonMultiInstanceBinaryMessengerWrapper

class NativeView(context: Context, viewId: Int, args: Any?, messenger: BinaryMessenger) : PlatformView, NativeTextApi {

    private val textView: TextView
    private val wrapper: LinearLayout

    private val flutterApi: FlutterTextApiHandler
    private val pigeonMessenger: PigeonMultiInstanceBinaryMessengerWrapper

    init {

        // wrapper view to maintain the correct sizing from native
        wrapper = LinearLayout(context)
        wrapper.id = viewId
        wrapper.setBackgroundColor(Color.GREEN)
        wrapper.layoutParams =  ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT)
        wrapper.gravity = Gravity.CENTER

        textView = TextView(context)
        textView.text = "NativeView_$viewId"

        wrapper.addView(textView)

        //setup pigeon
        pigeonMessenger = PigeonMultiInstanceBinaryMessengerWrapper(messenger, "id_$viewId");

        NativeTextApi.setUp(pigeonMessenger, this)
        flutterApi = FlutterTextApiHandler(pigeonMessenger)

    }


    override fun getView(): View? {
        return wrapper
    }

    override fun dispose() {
        //TODO("Not yet implemented")
    }

    //pigeon API
    override fun setText(text: String) {
        val finalText = "NativeView_$text"
        textView.text = finalText;
        flutterApi.textChanged(finalText, {})
    }


}

File: ios/Runner/NativeView.swift
//
//  NativeView.swift
//  Runner
//
//  Created by Daniel on 09/10/2023.
//

import Foundation
import Flutter
import UIKit
import SwiftUI

class NativeView: NSObject, FlutterPlatformView, NativeTextApi {

    private let _view: UIView
    private let _label: UILabel
    private var _flutterApi: FlutterTextApiHandler
    private let _pigeonMessenger: PigeonMultiInstanceBinaryMessengerWrapper

    init(
        frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?,
        binaryMessenger messenger: FlutterBinaryMessenger?
    ) {
        _view = UIView()
        _view.frame = frame
        _label = UILabel(frame: _view.bounds)

        //setup pigeon
        _pigeonMessenger = PigeonMultiInstanceBinaryMessengerWrapper(with: messenger!, channelSuffix: "id_\(viewId)")
        _flutterApi = FlutterTextApiHandler(binaryMessenger: _pigeonMessenger)
        super.init()

        // setup views
        _view.backgroundColor = UIColor.cyan

        _label.translatesAutoresizingMaskIntoConstraints = false
        _label.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        _label.textAlignment = .center
        
        _label.text = "NativeView_\(viewId)"

        _view.addSubview(_label)
        
        NativeTextApiSetup.setUp(binaryMessenger: _pigeonMessenger, api: self)
    }


    func view() -> UIView {
        return _view
    }
    
    func setText(text: String) throws {
        let finalText = "NativeView_\(text)"
        _label.text = finalText
        _flutterApi.textChanged(text: finalText, completion: {})
    }

}
